packages/sockets/src/logic/creators/createSocketCommunicator.ts

```diff
@@ -100,37 +119,87 @@ export default function createSocketCommunicator({
     }
   }
 
-  async function send(data: unknown, processes: Process[], abort?: Abort) {
+  async function send(data: unknown, process: Process, abort?: Abort) {
     return new Promise<void>((resolve, reject) => {
       function handleAbort() {
         reject(abort?.signal.reason);
       }
 
       abort?.signal.addEventListener('abort', handleAbort, { once: true });
 
-      processes.forEach((process) => {
+      if (areProcessesEqual(process, selfProcess)) {
+        selfQueue.enqueue({
+          data,
+          process: selfProcess,
+        });
+      } else {
         const connection = getConnectionByProcess(selfConnections, process);
 
         if (!connection) {
           throw new Error('Connection not found');
         }
-
         connection.socket.emit(SocketEvent.SEND, data);
-      });
+      }
 
       abort?.signal.removeEventListener('abort', handleAbort);
       resolve(undefined);
     });
   }
 
-  function broadcast(data: unknown, abort?: Abort) {
+  async function broadcast(data: unknown, abort?: Abort) {
     const processes = selfConnections.map(({ device: { process } }) => process);
 
-    return send(data, processes, abort);
+    await Promise.all(processes.map((process) => send(data, process, abort)));
+  }
+
+  async function scatter(
+    data: unknown[],
+    sendStartIndex: number,
+    sendCount: number,
+    buf: Array<ProcessWithData>,
+    recvStartIndex: number,
+    recvCount: number,
+    root: number,
+    abort?: Abort,
+  ) {
+    const isCalledByRoot = selfProcess.rank === root;
+    if (isCalledByRoot) {
+      // apply split and send to all processes
+      const allDevicesNumber = selfConnections.length + 1;
+      const sliced = data.slice(sendStartIndex);
+      const splitted = _.chunk(
```

Comment:
suggestion(non-blocking): i would move all used lodash methods into `utils` module with re-exports  to allow tree shaking in the future